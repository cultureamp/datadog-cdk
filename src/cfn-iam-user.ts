// Generated by cdk-import
import * as cdk from "@aws-cdk/core"

/* eslint-disable @typescript-eslint/naming-convention */

/**
 * Datadog Application User 1.2.0
 *
 * @schema CfnUserProps
 */
export interface CfnUserProps {
  /**
   * Credentials for the Datadog API
   *
   * @schema CfnUserProps#DatadogCredentials
   */
  readonly datadogCredentials: CfnUserPropsDatadogCredentials

  /**
   * The role of the user
   *
   * @schema CfnUserProps#AccessRole
   */
  readonly accessRole?: CfnUserPropsAccessRole

  /**
   * User email
   *
   * @schema CfnUserProps#Email
   */
  readonly email?: string

  /**
   * User handle (a valid email)
   *
   * @schema CfnUserProps#Handle
   */
  readonly handle: string

  /**
   * User name
   *
   * @schema CfnUserProps#Name
   */
  readonly name?: string
}

/**
 * Converts an object of type 'CfnUserProps' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
function toJson_CfnUserProps(
  obj: CfnUserProps | undefined
): Record<string, any> | undefined {
  if (obj === undefined) {
    return undefined
  }
  const result = {
    DatadogCredentials: toJson_CfnUserPropsDatadogCredentials(
      obj.datadogCredentials
    ),
    AccessRole: obj.accessRole,
    Email: obj.email,
    Handle: obj.handle,
    Name: obj.name,
  }
  // filter undefined values
  return Object.entries(result).reduce(
    (r, i) => (i[1] === undefined ? r : { ...r, [i[0]]: i[1] }),
    {}
  )
}
/* eslint-enable max-len, quote-props */

/**
 * Credentials for the Datadog API
 *
 * @schema CfnUserPropsDatadogCredentials
 */
export interface CfnUserPropsDatadogCredentials {
  /**
   * Datadog API key
   *
   * @schema CfnUserPropsDatadogCredentials#ApiKey
   */
  readonly apiKey: string

  /**
   * Datadog application key
   *
   * @schema CfnUserPropsDatadogCredentials#ApplicationKey
   */
  readonly applicationKey: string

  /**
   * Datadog API URL (defaults to https://api.datadoghq.com)
   *
   * @schema CfnUserPropsDatadogCredentials#ApiURL
   */
  readonly apiUrl?: string
}

/**
 * Converts an object of type 'CfnUserPropsDatadogCredentials' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
function toJson_CfnUserPropsDatadogCredentials(
  obj: CfnUserPropsDatadogCredentials | undefined
): Record<string, any> | undefined {
  if (obj === undefined) {
    return undefined
  }
  const result = {
    ApiKey: obj.apiKey,
    ApplicationKey: obj.applicationKey,
    ApiURL: obj.apiUrl,
  }
  // filter undefined values
  return Object.entries(result).reduce(
    (r, i) => (i[1] === undefined ? r : { ...r, [i[0]]: i[1] }),
    {}
  )
}
/* eslint-enable max-len, quote-props */

/**
 * The role of the user
 *
 * @schema CfnUserPropsAccessRole
 */
export enum CfnUserPropsAccessRole {
  /** adm */
  ADM = "adm",
  /** ro */
  RO = "ro",
  /** st */
  ST = "st",
}

/**
 * A CloudFormation `Datadog::IAM::User`
 *
 * @cloudformationResource Datadog::IAM::User
 * @stability external
 */
export class CfnUser extends cdk.CfnResource {
  /**
   * The CloudFormation resource type name for this resource class.
   */
  public static readonly CFN_RESOURCE_TYPE_NAME = "Datadog::IAM::User"

  /**
   * Resource props.
   */
  public readonly props: CfnUserProps

  /**
   * Attribute `Datadog::IAM::User.Disabled`
   */
  public readonly attrDisabled: cdk.IResolvable
  /**
   * Attribute `Datadog::IAM::User.Verified`
   */
  public readonly attrVerified: cdk.IResolvable

  /**
   * Create a new `Datadog::IAM::User`.
   *
   * @param scope - scope in which this resource is defined
   * @param id    - scoped id of the resource
   * @param props - resource properties
   */
  constructor(scope: cdk.Construct, id: string, props: CfnUserProps) {
    super(scope, id, {
      type: CfnUser.CFN_RESOURCE_TYPE_NAME,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- null return as parameter is not null
      properties: toJson_CfnUserProps(props)!,
    })

    this.props = props

    this.attrDisabled = this.getAtt("Disabled")
    this.attrVerified = this.getAtt("Verified")
  }
}
