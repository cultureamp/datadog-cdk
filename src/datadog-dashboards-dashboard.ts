// Generated by cdk-import
import * as cdk from "@aws-cdk/core"

/* eslint-disable @typescript-eslint/naming-convention */

/**
 * Datadog Dashboard 2.1.0
 *
 * @schema CfnDashboardProps
 */
export interface CfnDashboardProps {
  /**
   * JSON string of the dashboard definition
   *
   * @schema CfnDashboardProps#DashboardDefinition
   */
  readonly dashboardDefinition: string
}

/**
 * Converts an object of type 'CfnDashboardProps' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_CfnDashboardProps(
  obj: CfnDashboardProps | undefined
): Record<string, any> | undefined {
  if (obj === undefined) {
    return undefined
  }
  const result = {
    DashboardDefinition: obj.dashboardDefinition,
  }
  // filter undefined values
  return Object.entries(result).reduce(
    (r, i) => (i[1] === undefined ? r : { ...r, [i[0]]: i[1] }),
    {}
  )
}
/* eslint-enable max-len, quote-props */

/**
 * A CloudFormation `Datadog::Dashboards::Dashboard`
 *
 * @cloudformationResource Datadog::Dashboards::Dashboard
 * @stability external
 * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Dashboards-Dashboard
 */
export class CfnDashboard extends cdk.CfnResource {
  /**
   * The CloudFormation resource type name for this resource class.
   */
  public static readonly CFN_RESOURCE_TYPE_NAME =
    "Datadog::Dashboards::Dashboard"

  /**
   * Resource props.
   */
  public readonly props: CfnDashboardProps

  /**
   * Attribute `Datadog::Dashboards::Dashboard.Id`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Dashboards-Dashboard
   */
  public readonly attrId: string
  /**
   * Attribute `Datadog::Dashboards::Dashboard.Url`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Dashboards-Dashboard
   */
  public readonly attrUrl: string

  /**
   * Create a new `Datadog::Dashboards::Dashboard`.
   *
   * @param scope - scope in which this resource is defined
   * @param id    - scoped id of the resource
   * @param props - resource properties
   */
  constructor(scope: cdk.Construct, id: string, props: CfnDashboardProps) {
    super(scope, id, {
      type: CfnDashboard.CFN_RESOURCE_TYPE_NAME,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- null return as parameter is not null
      properties: toJson_CfnDashboardProps(props)!,
    })

    this.props = props

    this.attrId = cdk.Token.asString(this.getAtt("Id"))
    this.attrUrl = cdk.Token.asString(this.getAtt("Url"))
  }
}
