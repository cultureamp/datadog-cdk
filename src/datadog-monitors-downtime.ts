// Generated by cdk-import
import * as cdk from "@aws-cdk/core"

/* eslint-disable @typescript-eslint/naming-convention */

/**
 * Datadog Monitors Downtime 3.1.0
 *
 * @schema CfnDowntimeProps
 */
export interface CfnDowntimeProps {
  /**
   * Whether or not this downtime is disabled
   *
   * @schema CfnDowntimeProps#Disabled
   */
  readonly disabled?: boolean

  /**
   * POSIX timestamp to end the downtime. If not provided, the downtime is in effect indefinitely (i.e. until you cancel it).
   *
   * @schema CfnDowntimeProps#End
   */
  readonly end?: number

  /**
   * Message on the downtime
   *
   * @schema CfnDowntimeProps#Message
   */
  readonly message?: string

  /**
   * A single monitor to which the downtime applies. If not provided, the downtime applies to all monitors.
   *
   * @schema CfnDowntimeProps#MonitorId
   */
  readonly monitorId?: number

  /**
   * A comma-separated list of monitor tags, to which the downtime applies. The resulting downtime applies to monitors that match ALL provided monitor tags.
   *
   * @schema CfnDowntimeProps#MonitorTags
   */
  readonly monitorTags?: string[]

  /**
   * The scope(s) to which the downtime applies
   *
   * @schema CfnDowntimeProps#Scope
   */
  readonly scope: string[]

  /**
   * POSIX timestamp to start the downtime. If not provided, the downtime starts the moment it is created.
   *
   * @schema CfnDowntimeProps#Start
   */
  readonly start?: number

  /**
   * The timezone for the downtime
   *
   * @schema CfnDowntimeProps#Timezone
   */
  readonly timezone?: string
}

/**
 * Converts an object of type 'CfnDowntimeProps' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_CfnDowntimeProps(
  obj: CfnDowntimeProps | undefined
): Record<string, any> | undefined {
  if (obj === undefined) {
    return undefined
  }
  const result = {
    Disabled: obj.disabled,
    End: obj.end,
    Message: obj.message,
    MonitorId: obj.monitorId,
    MonitorTags: obj.monitorTags?.map(y => y),
    Scope: obj.scope?.map(y => y),
    Start: obj.start,
    Timezone: obj.timezone,
  }
  // filter undefined values
  return Object.entries(result).reduce(
    (r, i) => (i[1] === undefined ? r : { ...r, [i[0]]: i[1] }),
    {}
  )
}
/* eslint-enable max-len, quote-props */

/**
 * A CloudFormation `Datadog::Monitors::Downtime`
 *
 * @cloudformationResource Datadog::Monitors::Downtime
 * @stability external
 * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
 */
export class CfnDowntime extends cdk.CfnResource {
  /**
   * The CloudFormation resource type name for this resource class.
   */
  public static readonly CFN_RESOURCE_TYPE_NAME = "Datadog::Monitors::Downtime"

  /**
   * Resource props.
   */
  public readonly props: CfnDowntimeProps

  /**
   * Attribute `Datadog::Monitors::Downtime.Active`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
   */
  public readonly attrActive: cdk.IResolvable
  /**
   * Attribute `Datadog::Monitors::Downtime.Canceled`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
   */
  public readonly attrCanceled: number
  /**
   * Attribute `Datadog::Monitors::Downtime.CreatorId`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
   */
  public readonly attrCreatorId: number
  /**
   * Attribute `Datadog::Monitors::Downtime.DowntimeType`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
   */
  public readonly attrDowntimeType: number
  /**
   * Attribute `Datadog::Monitors::Downtime.Id`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
   */
  public readonly attrId: number
  /**
   * Attribute `Datadog::Monitors::Downtime.ParentId`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
   */
  public readonly attrParentId: number
  /**
   * Attribute `Datadog::Monitors::Downtime.UpdaterId`
   * @link arn:aws:cloudformation:us-west-2::type/resource/7171b96e5d207b947eb72ca9ce05247c246de623/Datadog-Monitors-Downtime
   */
  public readonly attrUpdaterId: number

  /**
   * Create a new `Datadog::Monitors::Downtime`.
   *
   * @param scope - scope in which this resource is defined
   * @param id    - scoped id of the resource
   * @param props - resource properties
   */
  constructor(scope: cdk.Construct, id: string, props: CfnDowntimeProps) {
    super(scope, id, {
      type: CfnDowntime.CFN_RESOURCE_TYPE_NAME,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- null return as parameter is not null
      properties: toJson_CfnDowntimeProps(props)!,
    })

    this.props = props

    this.attrActive = this.getAtt("Active")
    this.attrCanceled = cdk.Token.asNumber(this.getAtt("Canceled"))
    this.attrCreatorId = cdk.Token.asNumber(this.getAtt("CreatorId"))
    this.attrDowntimeType = cdk.Token.asNumber(this.getAtt("DowntimeType"))
    this.attrId = cdk.Token.asNumber(this.getAtt("Id"))
    this.attrParentId = cdk.Token.asNumber(this.getAtt("ParentId"))
    this.attrUpdaterId = cdk.Token.asNumber(this.getAtt("UpdaterId"))
  }
}
