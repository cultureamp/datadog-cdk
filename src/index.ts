export * from "./datadog-dashboards-dashboard"
export * from "./datadog-iam-user"
export * from "./datadog-monitors-downtime"
export * from "./datadog-monitors-monitor"
export * from "./datadog-slos-slo"
