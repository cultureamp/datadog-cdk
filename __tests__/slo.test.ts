import * as cdk from "aws-cdk-lib"
import { CfnSlo, CfnSloPropsType, ThresholdTimeframe } from "../src"
import { Template } from "aws-cdk-lib/assertions"

const testStack = () => {
  const app = new cdk.App()
  return new cdk.Stack(app, "TestStack")
}

describe("Given an SLO construct", () => {
  describe("When I supply default values", () => {
    const stack = testStack()

    new CfnSlo(stack, "TestSlo", {
      name: "TestSlo",
      thresholds: [
        {
          target: 100,
          timeframe: ThresholdTimeframe.VALUE_7D,
          targetDisplay: "98.0",
        },
      ],
      type: CfnSloPropsType.METRIC,
    })

    test("Then the template synthesises correctly", () => {
      const template = Template.fromStack(stack)

      expect(template).toMatchSnapshot()
    })
  })

  describe("When I supply optional values", () => {
    const stack = testStack()

    new CfnSlo(stack, "TestSlo", {
      name: "TestSlo",
      thresholds: [
        {
          target: 100,
          timeframe: ThresholdTimeframe.VALUE_7D,
          targetDisplay: "98.0",
        },
      ],
      type: CfnSloPropsType.METRIC,
      description: "A test SLO",
      groups: ["TestGroup"],
      tags: ["asset:testasset"],
      monitorIds: [0, 1],
      query: {
        numerator: "numerator query",
        denominator: "denominator query",
      },
    })

    test("Then the template synthesises correctly", () => {
      const template = Template.fromStack(stack)

      expect(template).toMatchSnapshot()
    })
  })

  describe("When I reference fields", () => {
    const stack = testStack()

    const slo = new CfnSlo(stack, "TestSlo", {
      name: "TestSlo",
      thresholds: [
        {
          target: 100,
          timeframe: ThresholdTimeframe.VALUE_7D,
          targetDisplay: "98.0",
        },
      ],
      type: CfnSloPropsType.METRIC,
    })

    new cdk.CfnResource(stack, "TestResource", {
      type: "Test::Resource",
      properties: {
        idref: slo.ref,
        id: slo.attrId,
        deleted: slo.attrDeleted,
      },
    })

    test("Then the references appear in the synthesised output", () => {
      const template = Template.fromStack(stack)
      template.hasResourceProperties("Test::Resource", {
        deleted: {
          "Fn::GetAtt": ["TestSlo", "Deleted"],
        },
        id: {
          "Fn::GetAtt": ["TestSlo", "Id"],
        },
        idref: {
          Ref: "TestSlo",
        },
      })
    })
  })
})
