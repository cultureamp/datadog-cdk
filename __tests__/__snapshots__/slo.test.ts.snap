// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an SLO construct When I supply default values Then the template synthesises correctly 1`] = `
{
  "Resources": {
    "TestSlo": {
      "Properties": {
        "Name": "TestSlo",
        "Thresholds": [
          {
            "Target": 100,
            "TargetDisplay": "98.0",
            "Timeframe": "7d",
          },
        ],
        "Type": "metric",
      },
      "Type": "Datadog::SLOs::SLO",
    },
  },
}
`;

exports[`Given an SLO construct When I supply optional values Then the template synthesises correctly 1`] = `
{
  "Resources": {
    "TestSlo": {
      "Properties": {
        "Description": "A test SLO",
        "Groups": [
          "TestGroup",
        ],
        "MonitorIds": [
          0,
          1,
        ],
        "Name": "TestSlo",
        "Query": {
          "Denominator": "denominator query",
          "Numerator": "numerator query",
        },
        "Tags": [
          "asset:testasset",
        ],
        "Thresholds": [
          {
            "Target": 100,
            "TargetDisplay": "98.0",
            "Timeframe": "7d",
          },
        ],
        "Type": "metric",
      },
      "Type": "Datadog::SLOs::SLO",
    },
  },
}
`;
